text/0000-move-pointer.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: `move_pointer`
+- Start Date: 2016-05-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce a new pointer type `&move` that logically owns the pointed-to data, but does not control the backing memory. Also introduces the DerefMove trait to provide access to such a pointer.
+
+# Motivation
+[motivation]: #motivation
+
+This provides an elegant solution to passing DSTs by value, allowing `Box<FnOnce>` to work. It also will allow other usecases where trait methods should take self by "value".
+
+# Detailed design
+[design]: #detailed-design
+
+- Add a new pointer type `&move T`
+ - This pointer will get the same lifetime rules as `&` and `&mut` (such that it cannot outlive the allocation)
+ - but, it will allow the value to become invalid before the allocation does.
+ - Deref coercions as per RFC #241 apply.
+ - Variables of type `&move T` will only allow mutating `T` if they themselves are `mut`
+- Add a new unary operation `&move <value>`. With a special case such that `&move |x| x` parses as a move closure, requiring parentheses to parse as an owned pointer to a closure.
+ - This precedence can be implemented by ignoring the `move` when parsing unary operators if it is followed by a `|` or `||` token.
+- Add a new operator trait `DerefMove` to allow smart pointer types to return owned pointers to their contained data. 
+ - A type that implements `DerefMove` cannot implement `Drop` (as `DerefMove` provides equivalent functinality)
+
+```rust
+trait DerefMove: DerefMut
+{
+    /// Return an owned pointer to inner data
+    fn deref_move(&mut self) -> &move Self::Target;
```

Comment:
A teaser:

"type state" in my parlance would allow for something like:
```rust
fn foo(arg: {'lifetime, before_type, during_type, after_type}, ...) -> ...;
```
e.g.
```rust
fn take_ref(self: {'a, T, MutBorrowed<T>, T}) -> &'a mut T;
fn take_mut_ref(self: {'a, T, Aliased<T>, T}) -> &'a T;
```
Now suppose there is some `Uninit<T>`, which has no destructor and can only be written too (in which case it becomes a `T`).
```rust
fn move_in(self: {'unused, Uninit<T>, Whatever, T}, T) -> ();
fn move_out(self: {'unused, T, Whatever, Uninit<T>}) -> T;
```
Now suppose that `T::Toggle = Uninit<T>`, and `Uninit<T>::Toggle = T`, so we have a way to go there and back:
```rust
fn take_move_ref<T>(self: {'a, T, MutBorrowed<T>, T::Toggle}) -> &'a move T;

// Type App
take_move_ref::<T::Toggle>: fn (self: {'a, T::Toggle, MutBorrowed<T::Toggle>, T::Toggle::Toggle}) -> &'a move T::Toggle;
// Eliminate ::Toggle::Toggle
take_move_ref::<T::Toggle>: fn (self: {'a, T::Toggle, MutBorrowed<T::Toggle>, T}) -> &'a move T::Toggle;

type OutPtr<'a, T> = & 'a move T::Toggle;
```
And low and behold we get out-pointers for free!